src/Cache/Server/Server.php

```diff
@@ -0,0 +1,576 @@
+
+		$item->sync_to_storage();
+
+		$this->mark_not_updating( $updating_key, $is_network );
+
+		$this->exit_on_update_end();
+	}
+
+	/**
+	 * @param string $namespace
+	 * @param string $key
+	 *
+	 * @return bool
+	 */
+	public function is_queued_for_update( string $namespace, string $key ): bool {
+
+		return ! empty( $this->spawn_queue[ $this->full_key( $namespace, $key ) ] );
+	}
+
+	/**
+	 * @param string $namespace
+	 * @param string $key
+	 *
+	 * @return string
+	 */
+	private function full_key( string $namespace, string $key ) {
+
+		return $this->factory->prefix() . $namespace . $key;
+	}
+
+	/**
+	 * Adds the actions that will cause item flushing.
+	 *
+	 * @param ItemLogic $logic
+	 * @param bool      $for_network
+	 *
+	 * @return Server
+	 */
+	private function do_register( ItemLogic $logic, bool $for_network ) {
+
+		if ( $this->update_request_key() || doing_action( 'shutdown' ) ) {
+			throw new \BadMethodCallException(
+				'Cache item logic registration is not possible during cache update requests.'
+			);
+		}
+
+		$namespace = $logic->namespace();
+		$key       = $logic->key();
+
+		$this->registered[ $this->full_key( $namespace, $key ) ] = [ $for_network, $logic ];
+
+		$actions = $logic->deleting_actions();
+
+		foreach ( $actions as $action ) {
+			add_action( $action, function () use ( $logic ) {
+
+				$this->flush( $logic->namespace(), $logic->key() );
+			} );
+		}
+
+		return $this;
+	}
+
+	/**
+	 * Check the HTTP request to see if it is a cache update request and if the eventually requested key is registered.
+	 * If so return it, otherwise return an empty string.
+	 *
+	 * @return string
+	 */
+	private function update_request_key(): string {
+
+		$method = $_SERVER['REQUEST_METHOD'] ?? '';
+		$key    = $_SERVER[ 'HTTP_' . self::HEADER_KEY ] ?? '';
+
+		return $key && strtoupper( $method ) === 'HEAD' ? $key : '';
+	}
+
+	/**
+	 * Use transients to mark the given key as currently being updated in a update request, to prevent
+	 * multiple concurrent updates.
+	 *
+	 * @param string $key
+	 * @param bool   $is_network
+	 *
+	 * @return bool
+	 */
+	private function mark_updating( string $key, bool $is_network ): bool {
+
+		$keys = $is_network
+			? get_site_transient( self::UPDATING_KEYS_TRANSIENT )
+			: get_transient( self::UPDATING_KEYS_TRANSIENT );
+
+		is_array( $keys ) or $keys = [];
+
+		$keys[] = $key;
+
+		return $is_network
+			? set_site_transient( self::UPDATING_KEYS_TRANSIENT, $keys )
+			: set_transient( self::UPDATING_KEYS_TRANSIENT, $keys );
+	}
+
+	/**
+	 * Remove the given key from transient storage to mark given key again available for updates.
+	 *
+	 * @param string $key
+	 * @param bool   $is_network
+	 *
+	 * @return bool
+	 */
+	private function mark_not_updating( string $key, bool $is_network ): bool {
+
+		$keys = $is_network
+			? get_site_transient( self::UPDATING_KEYS_TRANSIENT )
+			: get_transient( self::UPDATING_KEYS_TRANSIENT );
+
+		return $is_network
```

Comment:
I'd move the `(array)` cast to the retrieval above. And then I'd just `unset( $keys[ $key ] );`. This makes the `return` much more concise and more readable.